eth/db/header.py

```diff
@@ -296,6 +287,47 @@ def _persist_header_chain(
 
         return tuple(), tuple()
 
+    @classmethod
+    def _handle_gap_change(cls,
+                           db: DatabaseAPI,
+                           gap_change: GapChange,
+                           header: BlockHeaderAPI,
+                           genesis_parent_hash: Hash32) -> None:
+
+        if gap_change not in GAP_WRITES:
+            return
+
+        if gap_change is GapChange.GapFill:
+            expected_child = cls._get_canonical_head(db)
+            if header.hash != expected_child.parent_hash:
+                # We are trying to close a gap with an uncle. Reject!
+                raise ValidationError(f"{header} is not the parent of {expected_child}")
+
+        # We implicitly assert that persisted headers are canonical here.
+        # This assertion is made when persisting headers that are known to be part of a gap
+        # in the canonical chain.
+        # What we don't know is if this header will eventually lead up to the upper end of the
+        # gap (the checkpoint header) or if this is an alternative history. But we do ensure the
+        # integrity eventually. For one, we check the final header that would close the gap and if
+        # it does not match our expected child (the checkpoint header), we will reject it.
+        # Additionally, if we have persisted a chain of alternative history under the wrong
+        # assumption that it would be the canonical chain, but then at a later point it turns out it
+        # isn't and we overwrite it with the correct canonical chain, we make sure to
+        # recanonicalize all affected headers.
+        # IMPORTANT: Not only do we have to take this into account when we fill a gap, but also
+        # every time we write into a gap.
+        # Suppose we have a gap of 10 headers, then we fill 1 -5 with uncles from chain B.
+        # Now suppose we find the original chain A and attempt to write headers 1 - 10. At that
+        # point, we are under the assumption our current gap spans from just 6 - 10 because we have
+        # previously written headers 1 - 5 from chain B *believing* they are canonical which they
+        # turn out not to be. That means by the time we write headers 1 -5 again (but from chain A)
+        # we do not recognize it as writing to a known gap. Therefore by the time we write header 6,
+        # when we finally realize that we are attempting to fill in a gap, we have to backtrack
+        # the ancestors to add them to the canonical chain.
+
+        for ancestor in cls._find_new_ancestors(db, header, genesis_parent_hash):
+            cls._add_block_number_to_hash_lookup(db, ancestor)
```

Comment:
This will currently raise a `ParentNotFound` exception which I believe is the right thing to do. At least, everything becomes a bit simpler if we just continue with the assumption that headers must always be written sequentially. In other words, gaps aside, we can not write `A2` before `A1`  was written.

Reply:
Ah, right, that does make things simpler. Though I think most of the complexity is added back in with checkpoints. What about multiple checkpoints? Want to make sure re-canonicalization happens correctly (for example, if the checkpoint fills a gap).

```
A3 (set checkpoint)
B0:2 (now canonical)
A2 (set checkpoint)
<- what happens here?
A0:2
<- what about here?
```

Reply:
Oh you are bringing up an interesting case there! When we set `A2` as a new checkpoint it becomes the new canonical head and we stop caring about `A3`. I think that is the way it should be (For some reason, the user decided to set the checkpoint to `A2`, imho that should be given precedence).

But know what happens is that `A2` is closing a gap "backwards" which isn't what we want here! We get into a situation where our gap tracking would think we have a well-connected chain which isn't the case!

So, my idea is that `persist_checkpoint_header` must be altered to notice situations when it is backwards filling a gap and then do one of these things:

1. Raise an exception. Let client code now that they can not pick this checkpoint before they bring the database into a different state

2. Alternatively, remove the `B1` header to create a new gap to continue to track this missing connection. Removing just the single header before `A2` should be enough because we do not know where exactly the fork might be (which are uncles, which are actual parents)

Reply:
I will track this one with an [issue](https://github.com/ethereum/py-evm/issues/1928) and merge the PR as is